### ISSUE TITLE
composer install fails to build the autoloader, PHP Fatal error:  Uncaught TypeError: Argument 1 passed to Automattic\Jetpack\Autoloader\AutoloadGenerator::dump()

### ISSUE THREAD

**Issue by helgatheviking** [woocommerce/woocommerce#28622]
Me again with build script problem. 

**Prerequisites (mark completed items with an [x]):**
- [x] I have have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**
When pulling from the master branch, the autoloader needs to be rebuilt. So I run `composer install`. This is currently failing for me. 

**Expected behavior**
the autoloaders are created

**Actual behavior**
A fatal error that prevents the autoloaders from being created

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**
Steps to reproduce the bug:
pull latest master
open terminal
run `composer install`

**Screenshots**
```
helga@Lexa:/mnt/c/Users/helga/Local Sites/sandbox/app/public/wp-content/plugins/woocommerce$ composer install
A script named bin would override a Composer command and has been skipped
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Package operations: 0 installs, 4 updates, 0 removals
  - Updating automattic/jetpack-autoloader (v2.2.0 => v2.6.0): Downloading (100%)
  - Updating automattic/jetpack-constants (v1.5.0 => v1.5.1): Downloading (100%)
  - Updating woocommerce/woocommerce-admin (1.7.0 => 1.7.3): Downloading (100%)
  - Updating woocommerce/woocommerce-blocks (v3.8.0 => v3.8.1): Downloading (100%)
Generating autoload files
PHP Fatal error:  Uncaught TypeError: Argument 1 passed to Automattic\Jetpack\Autoloader\AutoloadGenerator::dump() must be an instance of Composer\Composer, instance of Composer\Config given, called in /mnt/c/Users/helga/Local Sites/sandbox/app/public/wp-content/plugins/woocommerce/vendor/automattic/jetpack-autoloader/src/CustomAutoloaderPlugin.php on line 123 and defined in /mnt/c/Users/helga/Local Sites/sandbox/app/public/wp-content/plugins/woocommerce/vendor/automattic/jetpack-autoloader/src/AutoloadGenerator.php:79
Stack trace:
#0 /mnt/c/Users/helga/Local Sites/sandbox/app/public/wp-content/plugins/woocommerce/vendor/automattic/jetpack-autoloader/src/CustomAutoloaderPlugin.php(123): Automattic\Jetpack\Autoloader\AutoloadGenerator->dump()
#1 [internal function]: Automattic\Jetpack\Autoloader\CustomAutoloaderPlugin->postAutoloadDump()
#2 phar:///mnt/c/ProgramData/ComposerSetup/bin/composer.phar/src/Composer/EventDispatcher/EventDispatcher.php(176): call_user_func()
#3 phar:///mnt/c/ProgramData/ComposerSetup/bin/composer.phar in /mnt/c/Users/helga/Local Sites/sandbox/app/public/wp-content/plugins/woocommerce/vendor/automattic/jetpack-autoloader/src/AutoloadGenerator.php on line 79

Fatal error: Uncaught TypeError: Argument 1 passed to Automattic\Jetpack\Autoloader\AutoloadGenerator::dump() must be an instance of Composer\Composer, instance of Composer\Config given, called in /mnt/c/Users/helga/Local Sites/sandbox/app/public/wp-content/plugins/woocommerce/vendor/automattic/jetpack-autoloader/src/CustomAutoloaderPlugin.php on line 123 and defined in /mnt/c/Users/helga/Local Sites/sandbox/app/public/wp-content/plugins/woocommerce/vendor/automattic/jetpack-autoloader/src/AutoloadGenerator.php:79
Stack trace:
#0 /mnt/c/Users/helga/Local Sites/sandbox/app/public/wp-content/plugins/woocommerce/vendor/automattic/jetpack-autoloader/src/CustomAutoloaderPlugin.php(123): Automattic\Jetpack\Autoloader\AutoloadGenerator->dump()
#1 [internal function]: Automattic\Jetpack\Autoloader\CustomAutoloaderPlugin->postAutoloadDump()
#2 phar:///mnt/c/ProgramData/ComposerSetup/bin/composer.phar/src/Composer/EventDispatcher/EventDispatcher.php(176): call_user_func()
#3 phar:///mnt/c/ProgramData/ComposerSetup/bin/composer.phar in /mnt/c/Users/helga/Local Sites/sandbox/app/public/wp-content/plugins/woocommerce/vendor/automattic/jetpack-autoloader/src/AutoloadGenerator.php on line 79
```

**Comment by rrennick** [comment]
@helgatheviking If you run `composer install` a second time immediately after the first does it work? If not, did you run `composer self-update`?

**Comment by helgatheviking** [comment]
@rrennick Hi Ron. I did need to update composer itself, so thanks for that. Though now I get a different issue.

```
Generating autoload files
> sh ./bin/package-update.sh
: not foundge-update.sh: 2:

: not foundge-update.sh: 16: }
: not foundge-update.sh: 17:
./bin/package-update.sh: 43: Syntax error: end of file unexpected (expecting "then")
Script sh ./bin/package-update.sh handling the post-install-cmd event returned with error code 2
```

and still can't finish generating the autoloader files.

**Comment by rrennick** [comment]
@helgatheviking Are you running on Windows? If so, there might be a difference in the `sh` implementation that is triggering the error. cc @Konamiman

**Comment by helgatheviking** [comment]
@rrennick Yes, but I'm running the Linux subsystem WSL. in theory, that's supposed to help me avoid quite a lot of issues where Windows commands are not the same as Linux/Mac commands. And it _has_, but still stuck on this one.

**Comment by rrennick** [comment]
@Konamiman Is on holiday break. He may be able to offer some insight on return.

**Comment by helgatheviking** [comment]
Can't say I'm working too hard myself. :)

**Comment by ObliviousHarmony** [comment]
This specific issue should be resolved by 4.9 as we have updated the autoloader to a version that fixes this.

Somewhat related: https://github.com/woocommerce/woocommerce/issues/28775 is a similar issue in that they both come as a result of executing the autoloader composer plugin transitively. I've opened a feature request so that we can prevent downstream packages from executing the plugin at all and avoid problems that arise.
